src/main/java/uk/gov/justice/hmpps/prison/service/AppointmentsService.kt

```diff
@@ -0,0 +1,461 @@
+import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderBookingRepository
+import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderIndividualScheduleRepository
+import uk.gov.justice.hmpps.prison.repository.jpa.repository.ReferenceCodeRepository
+import uk.gov.justice.hmpps.prison.repository.jpa.repository.ScheduledAppointmentRepository
+import uk.gov.justice.hmpps.prison.security.AuthenticationFacade
+import uk.gov.justice.hmpps.prison.service.support.ReferenceDomain
+import uk.gov.justice.hmpps.prison.service.support.StringWithAbbreviationsProcessor
+import uk.gov.justice.hmpps.prison.util.CalcDateRanges
+import java.time.Duration
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.util.Optional
+import java.util.stream.Collectors
+
+@Service
+@Validated
+@Transactional(readOnly = true)
+@Slf4j
+class AppointmentsService(
+  private val offenderBookingRepository: OffenderBookingRepository,
+  private val authenticationFacade: AuthenticationFacade,
+  private val locationService: LocationService,
+  private val referenceDomainService: ReferenceDomainService,
+  private val telemetryClient: TelemetryClient,
+  private val scheduledAppointmentRepository: ScheduledAppointmentRepository,
+  private val offenderIndividualScheduleRepository: OffenderIndividualScheduleRepository,
+  private val agencyLocationRepository: AgencyLocationRepository,
+  private val agencyInternalLocationRepository: AgencyInternalLocationRepository,
+  private val eventStatusRepository: ReferenceCodeRepository<EventStatus>,
+) {
+  @Transactional
+  fun createAppointments(@Valid appointments: AppointmentsToCreate): List<CreatedAppointmentDetails> {
+    assertThatRequestHasPermission(appointments)
+    assertFewerThanMaximumNumberOfBookingIds(appointments)
+
+    val defaults = appointments.appointmentDefaults
+
+    val appointmentLocation = agencyInternalLocationRepository.findByIdOrNull(defaults.locationId)
+      ?: throw EntityNotFoundException.withId(defaults.locationId)
+
+    val agencyId = findLocationInUserLocations(appointmentLocation)
+      .orElseThrow(BadRequestException.withMessage("Location does not exist or is not in your caseload."))
+      .agencyId
+
+    assertValidAppointmentType(defaults.appointmentType) // GETS event type
+
+    val flattenedDetails = appointments.withDefaults()
+
+    assertAdditionalAppointmentConstraints(flattenedDetails)
+
+    val appointmentsWithRepeats = withRepeats(appointments.repeat, flattenedDetails)
+    assertThatAppointmentsFallWithin(appointmentsWithRepeats, appointmentTimeLimit())
+
+    val eventStatus = eventStatusRepository.findByIdOrNull(EventStatus.SCHEDULED_APPROVED)
+      ?: throw RuntimeException("Internal error: Event status not recognised.")
+    val prison = agencyLocationRepository.findByIdOrNull(agencyId)
+      ?: throw RuntimeException("Internal error: Prison not found")
```

Comment:
Couldnt avoid getting the prison here as the agency id was obtained via some very old-school caseload sql